src/main/java/com/conveyal/datatools/manager/utils/MergeFeedUtils.java

```diff
@@ -0,0 +1,144 @@
+        if (csvReader == null) {
+            LOG.warn("Table {} not found in zip file: {}", table.name, zipFile.getName());
+            return ids;
+        }
+        Field[] fieldsFoundInZip = table.getFieldsFromFieldHeaders(csvReader.getHeaders(), null);
+        // Get the key field (id value) for each row.
+        int keyFieldIndex = getFieldIndex(fieldsFoundInZip, keyField);
+        while (csvReader.readRecord()) ids.add(csvReader.get(keyFieldIndex));
+        csvReader.close();
+        return ids;
+    }
+
+    public static String stopCodeFailureMessage(int stopsMissingStopCodeCount, int stopsCount, int specialStopsCount) {
+        return String.format(
+            "If stop_code is provided for some stops (for those with location_type = " +
+                "empty or 0), all stops must have stop_code values. The merge process " +
+                "found %d of %d total stops that were incorrectly missing stop_code values. " +
+                "Note: \"special\" stops with location_type > 0 need not specify this value " +
+                "(%d special stops found in feed).",
+            stopsMissingStopCodeCount,
+            stopsCount,
+            specialStopsCount
+        );
+    }
+
+    /**
+     * Collect zipFiles for each feed version before merging tables.
+     * Note: feed versions are sorted by first calendar date so that future dataset is iterated over first. This is
+     * required for the MTC merge strategy which prefers entities from the future dataset over past entities.
+     */
+    public static List<FeedToMerge> collectAndSortFeeds(Set<FeedVersion> feedVersions) {
+        return feedVersions.stream().map(version -> {
+            try {
+                return new FeedToMerge(version);
+            } catch (Exception e) {
+                LOG.error("Could not create zip file for version: {}", version.version);
+                return null;
+            }
+        }).filter(Objects::nonNull).filter(entry -> entry.version.validationResult != null
+            && entry.version.validationResult.firstCalendarDate != null)
+            // MTC-specific sort mentioned in above comment.
+            // TODO: If another merge strategy requires a different sort order, a merge type check should be added.
+            .sorted(Comparator.comparing(entry -> entry.version.validationResult.firstCalendarDate,
+                Comparator.reverseOrder())).collect(Collectors.toList());
+    }
+
+    /** Get the set of shared fields for all feeds being merged for a specific table. */
+    public static Set<Field> getSharedFields(List<FeedToMerge> feedsToMerge, Table table) throws IOException {
+        Set<Field> sharedFields = new HashSet<>();
+        // First, iterate over each feed to collect the shared fields that need to be output in the merged table.
+        for (FeedToMerge feed : feedsToMerge) {
+            CsvReader csvReader = table.getCsvReader(feed.zipFile, null);
```

Comment:
Please close this csvReader at some point.

Reply:
Handled in https://github.com/ibi-group/datatools-server/pull/376/commits/06b2156db917fa518ba09f157a55369277cc7aa7.